Geographic features - update error logging routines

lmgonzales opened the pull request:
Refactor code for error logging of coordinate errors during indexing. Also provide a configurable setting in import.properties to set the location where the error log is output to.

demiankatz commented on import/import.properties:
```diff
@@ -31,3 +31,6 @@ marc.include_errors = true
 # marc.unicode_normalize = C
+
+# coordinate error log location - where to put coordinate indexing error logs
+coordinate.log.path = /usr/local/vufind/import
```
Could we make this default to solrmarc.path if not set? Then we could leave it commented out by default and reduce the number of hard-coded paths necessary to get SolrMarc working. (But we'd still have the option to send the logs somewhere else when desired). Otherwise, I'm concerned that when people upgrade, they'll start getting errors because of a missing or incorrect setting here.

lmgonzales replied:
rewriting this so that coordinate.log.path is set in vufind.properties. Default is VUFIND_LOCAL_DIR/import and failing that VUFIND_HOME/import and failing that an error message will be output to indicate that error message will not be logged in a file.

demiankatz commented on import/index_java/src/org/vufind/index/GeoTools.java:
```diff
@@ -79,11 +87,10 @@
+                    } 
                 } else {
-                    logger.error(".......... Not indexing INVALID coordinates: [ {" + coords.get('d') + "} {" + coords.get('e') + "} {" + coords.get('f') + "} {" + coords.get('g') + "} ]");
+                    String msgError = "Poorly formed coordinate(s).";
```
Do we need to define a string here, or could we just pass the string directly to the logging function in this case?

lmgonzales replied:
Good point. Since the validateCoordinateValues logs the error, we can just send this message to the logging function here.

demiankatz commented on import/index_java/src/org/vufind/index/GeoTools.java:
```diff
@@ -106,14 +113,107 @@
+    * @return String logPath
+    */
+   public static String getErrorLogPath() {
+        String filename = "import.properties";
```
I don't think we want to hard-code import.properties here; the properties filename may be overridden as a parameter to the indexer. There may be an easier way to get all of these settings. I don't know the answer off the top of my head, but let me know if you need help digging.

lmgonzales replied:
I'm going to re-write this section and resubmit.

demiankatz commented on import/index_java/src/org/vufind/index/GeoTools.java:
```diff
@@ -280,9 +378,13 @@ protected boolean validateDDCoordinates(Record record, Double west, Double east,
-        String recNum = recID.getData();
-        logger.error("Record ID: " + recNum.trim() + " - Coordinates form a line at the pole");
+        String msgError = "Coordinates form a line at the pole";
+        HashMap<Character, String> coords = new HashMap();
```
This hashmap creation is repeated a lot of times. Maybe you need a helper method that just takes west, east, north, south and builds the map.

lmgonzales replied:
Great idea. I'll implement that. Thanks!

lmgonzales commented:
@demiankatz, I've gone ahead and updated the code based on our discussions so that the coordinate error log path is set in vufind.properties, and it fails over to VUFIND_LOCAL_DIR/import, and failing that to VUFIND_HOME/import. I also refactored the code and created a coordinate compiler helper method to create the HashMap for the coordinates. That streamlined things a lot. Thanks for the suggestions. Please let me know if you see any other places where the code can be improved.

demiankatz commented on import/index_java/src/org/vufind/index/GeoTools.java:
```diff
@@ -314,15 +402,15 @@ public boolean validateValues(Record record, Double west, Double east, Double no
+        logErrorMessage(record, coords, msgError);
         return false;
     }
     if (north > 90.0 || north < -90.0 || south > 90.0 || south < -90.0) {
```
Do we need two separate if statements here? It looks like the result of the condition is identical for both blocks. How about one big:

<pre>
if (west > 180.0 || west < -180.0 || east > 180.0 || east < -180.0
    || north > 90.0 || north < -90.0 || south > 90.0 || south < -90.0
) {
</pre>

instead? Using that formatting still keeps the two sets of conditions on their own lines, but it's a little more concise.

lmgonzales replied:
That's definitely more concise. Thanks!

demiankatz commented on import/vufind.properties:
```diff
@@ -6,3 +6,8 @@
+# coordinate error log location - where to put coordinate indexing error logs
+# if not set, the default path will be VUFIND_LOCAL_DIR/import, and failing that,
+# VUFIND_HOME/import.
+coordinate.log.path = /usr/local/vufind/import
```
Let's comment this out for now to use the default behavior by default.

lmgonzales replied:
Ok. Will update this.

lmgonzales commented on import/index_java/src/org/vufind/index/GeoTools.java:
```diff
@@ -46,10 +60,23 @@
     // Initialize logging category
     static Logger logger = Logger.getLogger(GeoTools.class.getName());
 
+    public GeoTools()
```
Yes. Good catch. Thanks!

lmgonzales commented on import/index_java/src/org/vufind/index/GeoTools.java:
```diff
@@ -106,14 +129,81 @@
     }
 
+    /**
+    * Log coordinate indexing errors to external log file.
```
Yes. Good point. Thanks for catching this.

lmgonzales commented on import/index_java/src/org/vufind/index/GeoTools.java:
```diff
@@ -106,14 +129,81 @@
+        // Initialize error logging variables
+        String msgError = message;
+        ControlField recID = (ControlField) record.getVariableField("001");
+        String recNum = recID.getData().trim();
```
Sounds good. I'll just sub in a "N/A" string for recNum if recID is null. That way we still can report out the coordinates at the very least even if we don't have a record ID to go with it. That should help to some extent with troubleshooting.

demiankatz commented:
Looks good, @lmgonzales -- I've gone ahead and merged this one. Thanks!
